File: app/screens/FeedbackScreen.js
import React, { useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Alert, ScrollView, Dimensions, KeyboardAvoidingView, Platform } from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import { RadioButton, TextInput } from 'react-native-paper';
import { LinearGradient } from 'expo-linear-gradient'; // Import LinearGradient

const { width, height } = Dimensions.get('window');

const FeedbackScreen = ({ navigation }) => {
  const [formData, setFormData] = useState({
    navigationEase: '',
    reminderHelpfulness: '',
    bluetoothPerformance: '',
    waterLevelAccuracy: '',
    hydrationUsefulness: '',
    recommendationFit: '',
    overallExperience: '',
    encountegreyBugs: '',
    additionalSuggestions: '',
  });

  const handleInputChange = (field, value) => {
    setFormData({
      ...formData,
      [field]: value,
    });
  };

  const handleSubmit = () => {
    const emptyFields = Object.values(formData).some(field => field === '');
    if (emptyFields) {
      Alert.alert('Incomplete Form', 'Please fill out all fields before submitting.');
      return;
    }

    console.log('Feedback saved successfully')
    Alert.alert('Feedback Submitted', 'Thank you for your feedback : )');
    navigation.goBack();
  };

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.container}
    >
      <LinearGradient
        colors={['#164a9e', '#0c3a85']} // Blue to white gradient
        style={styles.linearGradient}
      >
        <ScrollView contentContainerStyle={styles.scrollViewContent}>
          <TouchableOpacity onPress={() => navigation.goBack()} style={styles.backButton}>
            <Icon name="arrow-back" size={24} color="#fff" />
          </TouchableOpacity>

          <Text style={styles.title}>Feedback</Text>

          <Text style={styles.questionText}>1. Was the app easy to navigate and use?</Text>
          <RadioButton.Group
            onValueChange={(value) => handleInputChange('navigationEase', value)}
            value={formData.navigationEase}
          >
            <RadioButton.Item label="Very Easy" value="Very Easy" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Easy" value="Easy" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Neutral" value="Neutral" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Difficult" value="Difficult" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Very Difficult" value="Very Difficult" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
          </RadioButton.Group>

          <Text style={styles.questionText}>2. How helpful do you find the drink water reminders?</Text>
          <RadioButton.Group
            onValueChange={(value) => handleInputChange('reminderHelpfulness', value)}
            value={formData.reminderHelpfulness}
          >
            <RadioButton.Item label="Very Helpful" value="Very Helpful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Helpful" value="Helpful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Neutral" value="Neutral" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Unhelpful" value="Unhelpful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
            <RadioButton.Item label="Very Unhelpful" value="Very Unhelpful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
          </RadioButton.Group>

          <Text style={styles.questionText}>3. How well does the Bluetooth connectivity work?</Text>
        <RadioButton.Group
          onValueChange={(value) => handleInputChange('bluetoothPerformance', value)}
          value={formData.bluetoothPerformance}
        >
          <RadioButton.Item label="Excellent" value="Excellent" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Good" value="Good" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel} />
          <RadioButton.Item label="Average" value="Average" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Poor" value="Poor" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Very Poor" value="Very Poor" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
        </RadioButton.Group>

        <Text style={styles.questionText}>4. How accurate do you find the water level sensing feature?</Text>
        <RadioButton.Group
          onValueChange={(value) => handleInputChange('waterLevelAccuracy', value)}
          value={formData.waterLevelAccuracy}
        >
          <RadioButton.Item label="Very Accurate" value="Very Accurate" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Accurate" value="Accurate" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Neutral" value="Neutral" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Inaccurate" value="Inaccurate" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Very Inaccurate" value="Very Inaccurate" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
        </RadioButton.Group>

        <Text style={styles.questionText}>5. How useful is the hydration analysis feature for you?</Text>
        <RadioButton.Group
          onValueChange={(value) => handleInputChange('hydrationUsefulness', value)}
          value={formData.hydrationUsefulness}
        >
          <RadioButton.Item label="Very Useful" value="Very Useful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Useful" value="Useful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Neutral" value="Neutral" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Not Very Useful" value="Not Very Useful" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Not Useful at All" value="Not Useful at All" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
        </RadioButton.Group>

        <Text style={styles.questionText}>6. How well do the hydration recommendations fit your needs?</Text>
        <RadioButton.Group
          onValueChange={(value) => handleInputChange('recommendationFit', value)}
          value={formData.recommendationFit}
        >
          <RadioButton.Item label="Perfectly" value="Perfectly" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Well" value="Well" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Neutral" value="Neutral" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Poorly" value="Poorly" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Not at All" value="Not at All" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
        </RadioButton.Group>

        <Text style={styles.questionText}>7. Did you encounter any bugs or crashes while using the app?</Text>
        <TextInput
          mode="outlined"
          multiline
          numberOfLines={4}
          placeholder="Describe any bugs or crashes..."
          value={formData.encountegreyBugs}
          onChangeText={(value) => handleInputChange('encountegreyBugs', value)}
          style={styles.textInput}
        />

        <Text style={styles.questionText}>8. How would you rate your overall experience with the app?</Text>
        <RadioButton.Group
          onValueChange={(value) => handleInputChange('overallExperience', value)}
          value={formData.overallExperience}
        >
          <RadioButton.Item label="Excellent" value="Excellent" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Good" value="Good" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Average" value="Average" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Poor" value="Poor" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
          <RadioButton.Item label="Very Poor" value="Very Poor" uncheckedColor="grey" color="yellow" labelStyle={styles.radioLabel}/>
        </RadioButton.Group>

        <Text style={styles.questionText}>9. Any Additional Suggestions for this app?</Text>
        <TextInput
          mode="outlined"
          multiline
          numberOfLines={4}
          placeholder="Share your suggestions..."
          value={formData.additionalSuggestions}
          onChangeText={(value) => handleInputChange('additionalSuggestions', value)}
          style={styles.textInput}
        />

          <TouchableOpacity style={styles.submitButton} onPress={handleSubmit}>
            <Text style={styles.submitButtonText}>Submit</Text>
          </TouchableOpacity>
        </ScrollView>
      </LinearGradient>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  linearGradient: {
    flex: 1,
    padding: width * 0.04,
  },
  backButton: {
    position: 'absolute',
    left: width * 0.04,
    top: height * 0.02,
  },
  title: {
    fontSize: width * 0.08,
    fontWeight: 'bold',
    textAlign: 'center',
    marginVertical: height * 0.05,
    color: '#fff', // Title color
  },
  scrollViewContent: {
    paddingBottom: height * 0.05,
  },
  questionText: {
    fontSize: width * 0.04,
    fontWeight: 'bold',
    color: '#fff', // White question text
    marginVertical: height * 0.015,
  },
  radioLabel: {
    color: '#fff', // White radio button labels
  },
  textInput: {
    marginBottom: height * 0.02,
    backgroundColor: '#cacfce',
  },
  submitButton: {
    backgroundColor: 'green',
    paddingVertical: height * 0.015,
    borderRadius: 8,
    alignItems: 'center',
    marginTop: height * 0.02,
  },
  submitButtonText: {
    color: '#fff',
    fontSize: width * 0.045,
    fontWeight: 'bold',
  },
});

export default FeedbackScreen;

File: flutter/lib/feedback_screen.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'dart:convert';

class FeedbackScreen extends StatefulWidget {
  @override
  _FeedbackScreenState createState() => _FeedbackScreenState();
}

class _FeedbackScreenState extends State<FeedbackScreen> {
  final FocusNode _focusNodeBugs = FocusNode();
  final FocusNode _focusNodeSuggestions = FocusNode();
  TextEditingController _encounteredBugsController = TextEditingController();
  TextEditingController _additionalSuggestionsController =
      TextEditingController();

  // Secure Storage
  final FlutterSecureStorage _secureStorage = FlutterSecureStorage();

  Map<String, String?> formData = {
    'navigationEase': null,
    'reminderHelpfulness': null,
    'bluetoothPerformance': null,
    'waterLevelAccuracy': null,
    'hydrationUsefulness': null,
    'recommendationFit': null,
    'overallExperience': null,
    'encounteredBugs': '',
    'additionalSuggestions': '',
  };

  void handleInputChange(String field, String? value) {
    setState(() {
      formData[field] = value?.trim();
    });
  }

  Future<void> handleSubmit() async {
    // Show dialog if form is incomplete
    if (formData.values.any((field) => field == null || field.isEmpty)) {
      showDialog(
        context: context,
        builder: (context) => AlertDialog(
          title: Text('Incomplete Form'),
          content: Text('Please fill out all fields before submitting.'),
          actions: [
            TextButton(
              onPressed: () => Navigator.pop(context),
              child: Text('OK'),
            ),
          ],
        ),
      );
      return;
    }

    // Show loading dialog
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        content: Row(
          children: [
            CircularProgressIndicator(),
            SizedBox(width: 20),
            Text("Submitting Feedback..."),
          ],
        ),
      ),
    );

    try {
      // Retrieve JWT token from secure storage
      String? token = await _secureStorage.read(key: 'access_token');
      if (token == null) {
        Navigator.pop(context); // Close loading dialog if token is null
        _showErrorDialog("Authentication required");
        return;
      }

      // Prepare the data
      Map<String, String?> data = {...formData};
      data['userToken'] = token;

      // Send feedback to backend
      final response = await http.post(
        Uri.parse(
            'http://192.168.149.8:8081/feedbacks'), // Replace with actual backend URL
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token', // Use the token for authorization
        },
        body: json.encode({
          'navigationEase': formData['navigationEase'],
          'reminderHelpfulness': formData['reminderHelpfulness'],
          'bluetoothPerformance': formData['bluetoothPerformance'],
          'waterLevelAccuracy': formData['waterLevelAccuracy'],
          'hydrationUsefulness': formData['hydrationUsefulness'],
          'recommendationFit': formData['recommendationFit'],
          'overallExperience': formData['overallExperience'],
          'encounteredBugs': _encounteredBugsController.text,
          'additionalSuggestions': _additionalSuggestionsController.text,
        }),
      );

      if (response.statusCode == 200) {
        Navigator.pop(context); // Close loading dialog
        _showSuccessDialog();
      } else {
        Navigator.pop(context); // Close loading dialog
        _showErrorDialog("Failed to submit feedback. Please try again.");
      }
    } catch (e) {
      Navigator.pop(context); // Close loading dialog
      _showErrorDialog("An error occurred. Please try again.");
    }
  }

  void _showSuccessDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Feedback Submitted'),
        content: Text('Thank you for your feedback!'),
        actions: [
          TextButton(
            onPressed: () {
              setState(() {
                formData = {
                  'navigationEase': null,
                  'reminderHelpfulness': null,
                  'bluetoothPerformance': null,
                  'waterLevelAccuracy': null,
                  'hydrationUsefulness': null,
                  'recommendationFit': null,
                  'overallExperience': null,
                  'encounteredBugs': '',
                  'additionalSuggestions': '',
                };
                _encounteredBugsController.clear();
                _additionalSuggestionsController.clear();
              });
              Navigator.pop(context);
              _focusNodeBugs.unfocus();
              _focusNodeSuggestions.unfocus();
            },
            child: Text('OK'),
          ),
        ],
      ),
    );
  }

  void _showErrorDialog(String message) {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Error'),
        content: Text(message),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('OK'),
          ),
        ],
      ),
    );
  }

  @override
  void dispose() {
    _encounteredBugsController.dispose();
    _additionalSuggestionsController.dispose();
    _focusNodeBugs.dispose();
    _focusNodeSuggestions.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blue[900],
      appBar: AppBar(
        backgroundColor: Colors.blue[900],
        leading: IconButton(
          icon: Icon(Icons.arrow_back),
          onPressed: () => Navigator.pop(context),
        ),
      ),
      body: SingleChildScrollView(
        padding: EdgeInsets.all(20),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Container(
              padding: EdgeInsets.only(top: 20),
              child: Center(
                child: Text(
                  'Feedback',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 30,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
            SizedBox(height: 40),
            _buildQuestion(
                '1. Was the app easy to navigate and use?', 'navigationEase'),
            _buildQuestion(
                '2. How helpful do you find the drink water reminders?',
                'reminderHelpfulness'),
            _buildQuestion('3. How well does the Bluetooth connectivity work?',
                'bluetoothPerformance'),
            _buildQuestion(
                '4. How accurate do you find the water level sensing feature?',
                'waterLevelAccuracy'),
            _buildQuestion(
                '5. How useful is the hydration analysis feature for you?',
                'hydrationUsefulness'),
            _buildQuestion(
                '6. How well do the hydration recommendations fit your needs?',
                'recommendationFit'),
            _buildQuestion(
                '7. How would you rate your overall experience with the app?',
                'overallExperience'),
            SizedBox(height: 20),
            _buildTextField(
                '8. Did you encounter any bugs or crashes while using the app?',
                'encounteredBugs',
                _encounteredBugsController,
                _focusNodeBugs),
            _buildTextField(
                '9. Any additional suggestions for improvement?',
                'additionalSuggestions',
                _additionalSuggestionsController,
                _focusNodeSuggestions),
            SizedBox(height: 20),
            Center(
              child: ElevatedButton(
                onPressed: handleSubmit,
                child: Text(
                  'Submit',
                  style: TextStyle(fontWeight: FontWeight.bold, fontSize: 18),
                ),
                style: ElevatedButton.styleFrom(
                  backgroundColor: Colors.green,
                  foregroundColor: Colors.white,
                  padding: EdgeInsets.symmetric(vertical: 20, horizontal: 40),
                  shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(5)),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildQuestion(String questionText, String fieldName) {
    return Padding(
      padding: EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(questionText,
              style: TextStyle(color: Colors.white, fontSize: 16)),
          Column(
            children: [
              _buildRadioOption(fieldName, 'Very Easy'),
              _buildRadioOption(fieldName, 'Easy'),
              _buildRadioOption(fieldName, 'Neutral'),
              _buildRadioOption(fieldName, 'Difficult'),
              _buildRadioOption(fieldName, 'Very Difficult'),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildRadioOption(String fieldName, String label) {
    return RadioListTile<String>(
      title: Text(label, style: TextStyle(color: Colors.white)),
      value: label,
      groupValue: formData[fieldName],
      onChanged: (value) {
        handleInputChange(fieldName, value);
      },
      activeColor: Colors.yellow,
    );
  }

  Widget _buildTextField(String label, String fieldName,
      TextEditingController controller, FocusNode focusNode) {
    return Padding(
      padding: EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label, style: TextStyle(color: Colors.white, fontSize: 16)),
          TextField(
            focusNode: focusNode,
            controller: controller,
            onChanged: (value) => handleInputChange(fieldName, value),
            maxLines: 4,
            decoration: InputDecoration(
              hintText: 'Type here...',
              hintStyle: TextStyle(color: Colors.grey),
              filled: true,
              fillColor: Colors.white,
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(5),
                borderSide: BorderSide.none,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

void main() {
  runApp(MaterialApp(
    home: FeedbackScreen(),
  ));
}
